From 2960508b33179551bcbb842693ee8999044ba04f Mon Sep 17 00:00:00 2001
From: Chris Friedt <cfriedt@tenstorrent.com>
Date: Sat, 15 Nov 2025 16:51:03 -0500
Subject: [PATCH] libc: minimal: add missing ctype.h functions

Add the functions below to the minimal libc ctype.h since they are
missing, and are required as of C89 (C99 for `isblank()`)

* `isblank()`
* `islower()`
* `ispunct()`

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>
---
 lib/libc/minimal/include/ctype.h | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/lib/libc/minimal/include/ctype.h b/lib/libc/minimal/include/ctype.h
index f9a4bdf7b4c47..78036b8333b9e 100644
--- a/lib/libc/minimal/include/ctype.h
+++ b/lib/libc/minimal/include/ctype.h
@@ -23,6 +23,11 @@ static inline int isalpha(int c)
 	return (int)((((unsigned)c|32u)-(unsigned)'a') < 26U);
 }
 
+static inline int isblank(int c)
+{
+	return (int)((c == (int)' ') || (c == (int)'\t'));
+}
+
 static inline int isspace(int c)
 {
 	return (int)(c == (int)' ' || ((unsigned)c-(unsigned)'\t') < 5U);
@@ -45,6 +50,11 @@ static inline int isdigit(int a)
 	return (int)(((unsigned)(a)-(unsigned)'0') < 10U);
 }
 
+static inline int islower(int c)
+{
+	return (int)(((unsigned)(c) - (unsigned)'a') < 26U);
+}
+
 static inline int isxdigit(int a)
 {
 	unsigned int ua = (unsigned int)a;
@@ -69,6 +79,11 @@ static inline int isalnum(int chr)
 	return (int)(isalpha(chr) || isdigit(chr));
 }
 
+static inline int ispunct(int c)
+{
+	return (int)(isgraph(c) && !isalnum(c));
+}
+
 static inline int iscntrl(int c)
 {
 	return (int)((((unsigned int)c) <= 31U) || (((unsigned int)c) == 127U));
